Fix for refresh of SUI $state.FeatureNav issue

chalettu opened the pull request:
Fix for Pivotal ticket https://www.pivotaltracker.com/story/show/138056447.

chalettu commented:
@miq-bot add_label bug
@miq-bot add_label euwe/no

AllenBW commented on client/app/core/layouts.config.js:
```diff
@@ -19,25 +19,31 @@ function getLayouts() {
 
 /** @ngInject */
-function enterApplication(Polling, lodash, $state, NavCounts, Navigation) {
+function enterApplication(Polling, lodash, $state, NavCounts, Navigation, Session, $timeout) {
```
Unused `Session` can be removed

AllenBW commented on client/app/core/layouts.config.js:
```diff
@@ -19,25 +19,31 @@ function getLayouts() {
+  function initBadgeCount() {
+    if (angular.isDefined($state.navFeatures)) {
+      angular.forEach(NavCounts.counts, updateCount);
+      angular.forEach(Navigation.items, function (value, key) {
```
Unless you're using [arrow shorthand](http://babeljs.io/learn-es2015/#ecmascript-2015-features-arrows-and-lexical-this)

prefer use of named functions, similar to the way `updateCount` is used on line 27

chalettu commented:
@miq-bot add_label wip

AllenBW commented on client/app/components/services/service-explorer/service-explorer.component.js:
```diff
@@ -40,7 +40,7 @@ function ComponentController($state, ServicesState, Language, ListView, Chargeba
       cardConfig: getCardConfig(),
       listConfig: getListConfig(),
-      listActions: getListActions(),
+      listActions: {},
```
I believe this to be a symptom the problem, not the underlying issue.  A fix shouldn't mess with this component.

AllenBW commented on client/app/core/session.service.js:
```diff
@@ -87,42 +91,56 @@ export function SessionFactory($http, $sessionStorage, gettextCatalog, $window,
-    $state.actionFeatures = setRBACForActions(productFeatures);
-    $state.navFeatures = setRBACForNavigation(productFeatures);
+
+    setRBACForActions(productFeatures).then(function (actionsRBAC) {
```
Furthering the feature bleed of RBAC functionality isn't ideal (that lil fixme is on to something)

chalettu commented:
@miq-bot remove_label wip

chalettu commented:
@AllenBW Take a look.  This PR is finally ready for merging

AllenBW commented on client/app/core/session.service.js:
```diff
@@ -83,99 +99,10 @@ export function SessionFactory($http, $sessionStorage, gettextCatalog, $window,
-      if (angular.isDefined(productFeatures[actions[i]])) {
-        return true;
-      }
+    if (model.token !== null) {
```
`return angular.isObject(model.token)? model.token : false;`
will that work here? (trade five lines for one)

jjlangholtz commented on client/app/core/rbac.service.js:
```diff
@@ -21,4 +30,101 @@ export function RBACFactory() {
+  }
+
+  function isAnyActionAllowed(actions, productFeatures) {
+    for (var i = 0; i <= actions.length; i++) {
```
Doesn't matter all too much since `actions` is a small array but this is a job for `Array#some`.

```javascript
return actions.some(action => angular.isDefined(productFeatures[action]);
```

That way the function always returns a boolean and iteration will stop as soon as possible.

AllenBW replied:
Even less! though eslint will holla about `action` not being in parenthesis

jjlangholtz commented on client/app/core/authorization.config.js:
```diff
@@ -61,16 +65,18 @@ export function authInit($rootScope, $state, Session, $sessionStorage, logger, L
+    syncSession().then(rbacReloadOrLogin(toState, toParams));
+  }
+  function syncSession() {
+    var deferred = $q.defer();
```
For new code I think we should avoid using the `$q` service and use the native ES6 `Promise` instead.

The same 3 lines in this function using `$q` could be rewritten as `.then(Promise.resolve())` in the last part of the `then` chain.

Also, after looking this over I think you can avoid adding the extra `.then` step and instead return the Promise chain directly:

```javascript
return Session.loadUser()
  .then(Language.onReload)
  .then(ServerInfo.set)
  .then(ProductInfo.set)
  .catch(badUser);
```

In either case there is no longer any `.catch` for `rbacReloadOrLogin` where there was before.

chalettu commented on client/app/core/authorization.config.js:
```diff
@@ -80,13 +86,16 @@ export function authInit($rootScope, $state, Session, $sessionStorage, logger, L
       .then(ServerInfo.set)
       .then(ProductInfo.set)
-      .then(rbacReloadOrLogin(toState, toParams))
+      .then(function() {
```
Since this functionality is resolving the promise, I can't move this functionality out to another function since that would be out of scope.

chalettu commented on client/app/core/layouts.config.js:
```diff
@@ -19,26 +19,30 @@ function getLayouts() {
+      actionFeatures[secondaryKey] = lodash.merge(secondaryValue, actionFeatures[secondaryKey]);
     });
   });
+  RBAC.setNavFeatures(navFeatures);
```
Are you saying you want me to change it to allow setting one feature at a time rather than all at once?

chalettu replied:
@AllenBW

chalettu commented:
@AllenBW I addressed all your feedback  but your last comment about moving RBAC.setNavFeatures in layout.config.js.  I can't move those functions up because they would be inside foreach statements that need to complete.

AllenBW commented:
ah! I see they are already in the function, ok no worries then. @chalettu

AllenBW commented on client/app/core/authorization.config.js:
```diff
@@ -45,11 +45,15 @@ export function authConfig($httpProvider) {
 
 /** @ngInject */
-export function authInit($rootScope, $state, Session, $sessionStorage, logger, Language, ServerInfo, ProductInfo, $window) {
+export function authInit($rootScope, $state, Session, $sessionStorage, logger, Language, ServerInfo, ProductInfo, $window, $q, RBAC) {
```
can now kill `$q` <3

AllenBW commented on client/app/core/rbac.service.js:
```diff
@@ -21,4 +32,91 @@ export function RBACFactory() {
+      administration: {show: entitledForService(productFeatures)},
+    };
+
+    return features;
```
Still returning on a setter, unless this isn't a setter, in which case it should be renamed.

AllenBW commented on client/app/core/rbac.service.js:
```diff
@@ -21,4 +32,91 @@ export function RBACFactory() {
+      serviceOwnership: {show: angular.isDefined(productFeatures.service_ownership)},
+    };
+
+    return features;
```
Still returning on a setter, unless this isn't a setter, in which case it should be renamed.
